src/OrleansRuntime/GrainDirectory/LocalGrainDirectory.cs

```diff
@@ -469,175 +472,235 @@ public SiloAddress CalculateTargetSilo(GrainId grain, bool excludeThisSiloIfStop
+                return null;
             }
-            else
-            {
-                RegistrationsSingleActRemoteSent.Increment();
-                // otherwise, notify the owner
-                Tuple<ActivationAddress, int> returnedAddress = await GetDirectoryReference(owner).RegisterSingleActivation(address, NUM_RETRIES);
-
-                // Caching optimization: 
-                // cache the result of a successfull RegisterSingleActivation call, only if it is not a duplicate activation.
-                // this way next local lookup will find this ActivationAddress in the cache and we will save a full lookup!
-                if (returnedAddress == null || returnedAddress.Item1 == null) return null;
-
-                if (!address.Equals(returnedAddress.Item1) || !IsValidSilo(address.Silo)) return returnedAddress.Item1;
 
-                var cached = new List<Tuple<SiloAddress, ActivationId>>(1) { Tuple.Create(address.Silo, address.Activation) };
-                // update the cache so next local lookup will find this ActivationAddress in the cache and we will save full lookup.
-                DirectoryCache.AddOrUpdate(address.Grain, cached, returnedAddress.Item2);
-                return returnedAddress.Item1;
+            if (hopCount >= HOP_LIMIT)
+            {
+                // we are not forwarding because there were too many hops already
+                throw new OrleansException(string.Format("Silo {0} is not owner of {1}, cannot forward {2} to owner {3} because hop limit is reached", MyAddress, grainId, operationDescription, owner));
             }
+
+            // forward to the silo that we think is the owner
+            return owner;
         }
 
-        public async Task RegisterAsync(ActivationAddress address)
+
+        public async Task<AddressAndTag> RegisterAsync(ActivationAddress address, bool singleActivation, int hopCount)
         {
-            registrationsIssued.Increment();
-            SiloAddress owner = CalculateTargetSilo(address.Grain);
-            if (owner == null)
+            var counterStatistic = 
+                singleActivation 
+                ? (hopCount > 0 ? this.RegistrationsSingleActRemoteReceived : this.registrationsSingleActIssued)
+                : (hopCount > 0 ? this.RegistrationsRemoteReceived : this.registrationsIssued);
+
+            counterStatistic.Increment();
+
+            // see if the owner is somewhere else (returns null if we are owner)
+            var forwardAddress = this.CheckIfShouldForward(address.Grain, hopCount, "RegisterAsync");
+
+            // on all silos other than first, we insert a retry delay and recheck owner before forwarding
+            if (hopCount > 0 && forwardAddress != null)
             {
-                // We don't know about any other silos, and we're stopping, so throw
-                throw new InvalidOperationException("Grain directory is stopping");
+                await Task.Delay(RETRY_DELAY);
```

Comment:
if we get to this point, it means that we were called from some other silo that thought we were the owner (because hopcount > 0, this method corresponds to the code that used to be in  RemoteGrainDirectory), but then it turns out that we are not. The name and the code are the same as before (we keep everything the same but remove code duplication).

As far as I understand the algorithm, waiting makes sense because the grain directory is unstable, and it is quite possible that after the wait, the local view of the cluster may have been updated and it may turn out we are actually the correct destination and do not need to forward after all.